lib/lint.ml

```diff
@@ -193,19 +300,100 @@ module Checks = struct
-    [
-      check_name_field;
-      check_version_field;
-      check_dune_subst;
-      check_dune_constraints;
-      check_checksums;
-      check_no_pin_depends;
-      check_no_extra_files;
-    ]
-
-  let run_checks ~pkg opam =
-    checks |> List.map (fun f -> f ~pkg opam) |> List.concat
+  let opam_lint ~pkg opam =
+    OpamFileTools.lint ~check_upstream:true opam
+    |> List.map (fun x -> (pkg, OpamLint x))
+
+  let is_perm_correct file =
+    Unix.stat file |> function
+    | { st_kind = S_REG; st_perm = 0o644; _ } -> true
+    | { st_kind = S_REG; st_perm = 0o664; _ } -> true
+    | _ -> false
+
+  let check_package_dir ~repo_dir ~pkg _opam =
+    let dir = path_from_pkg ~repo_dir pkg in
+    let check_file = function
+      | "opam" ->
+          let path = dir // "opam" in
+          if is_perm_correct path then [] else [ (pkg, ForbiddenPerm path) ]
+      | file -> [ (pkg, UnexpectedFile file) ]
+    in
+    get_files dir |> List.map check_file |> List.concat
+
+  (** [package_name_collision p0 p1] returns true if [p0] is similar to [p1].
+    Similarity is defined to be either:
+
+    - Case-insensitive string equality considering underscores ([_])
+      and dashes ([-]) to be equal
+    - A Levenshtein distance within 1/6 of the length of the string (rounding up),
+      with names of three characters or less ignored as a special case
+
+    As examples, by this relation:
+
+    - [lru-cache] and [lru_cache] collide
+    - [lru-cache] and [LRU-cache] collide
+    - [lru-cache] and [cache-lru] do not collide
+    - [ocaml] and [pcaml] collide
+    - [ocamlfind] and [ocamlbind] do not collide *)
+  let package_name_collision p0 p1 =
+    let dash_underscore p0 p1 =
+      let f = function '_' -> '-' | c -> c in
+      let p0 = String.map f p0 in
+      let p1 = String.map f p1 in
+      String.equal p0 p1
+    in
+    let levenstein_distance p0 p1 =
+      let l = String.length p0 in
+      if l <= 3 then false
+      else
+        let k = ((l - 1) / 16) + 2 in
+        (* Ignore distances of 1, too many false positives:
+           https://github.com/ocaml/opam-repository/pull/25678 *)
+        match Mula.Strings.Lev.get_distance ~k p0 p1 with
+        | None -> false
+        | Some n when n <= 1 -> false
+        | Some _ -> true
+    in
+    dash_underscore p0 p1 || levenstein_distance p0 p1
+
+  let check_name_collisions ~pkg packages _opam =
+    let pkg_name = pkg.OpamPackage.name |> OpamPackage.Name.to_string in
+    let pkg_name_lower = String.lowercase_ascii pkg_name in
+    let other_pkgs =
+      List.filter (fun s -> not @@ String.equal s pkg_name) packages
+    in
+    List.filter_map
+      (fun other_pkg ->
+        let other_pkg_lower = String.lowercase_ascii other_pkg in
+        if package_name_collision pkg_name_lower other_pkg_lower then
+          Some (pkg, NameCollision other_pkg)
+        else None)
+      other_pkgs
+
+  let checks ~newly_published ~repo_dir packages =
+    let new_checks =
```

Comment:
When reading this first thought "why have a special variable for newly added checks?". A name like this would have prevented my confusion (tho it seems silly in retrospect):
 
```suggestion
    let newly_published_checks =
```